hibernate-core/src/main/java/org/hibernate/engine/jdbc/batch/internal/AbstractBatchImpl.java

```diff
@@ -187,12 +187,17 @@ protected final void notifyObserversImplicitExecution() {
 		}
 	}
 
+	protected int getBatchPosition(){
+		return 0;
+	}
+
+
 	@Override
 	public void release() {
-		if ( getStatements() != null && !getStatements().isEmpty() ) {
+		if ( getStatements() != null && !getStatements().isEmpty() && getBatchPosition() != 0 ) {
 			LOG.batchContainedStatementsOnRelease();
+			releaseStatements();
 		}
-		releaseStatements();
 		observers.clear();
```

Comment:
I've pushed a [branch](https://github.com/gbadner/hibernate-core/tree/HHH-13050_naros_2) that includes a test that @dreab8 created to show that there is no statement leak.

I've added a [commit](https://github.com/gbadner/hibernate-core/commit/8c29a448d04ee37af7647ac31c139f9abb5d10c7) that adds a test with a custom `Batch` implementation similar to what @dreab8 did in his test.  In this new test, the custom batch throws an exception when `#addToBatch` is called. `#releaseStatements` does not get called because the batch gets aborted with `batchPosition` == 0. As a result the batch will still contain a `PreparedStatement` after `#release` is called. The test confirms that it is closed.